File: swift/TestDLTHM.swift
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* test driver and function exerciser for KYBER API Functions */

import Foundation
import core // comment out for Xcode

public func TestDLTHM(){
        let LOOPS=100
        var RAW=[UInt8](repeating: 0,count:100)
        var TAU=[UInt8](repeating: 0,count:32)
        var SK=[UInt8](repeating:0,count:DILITHIUM.SK_SIZE_3)
        var PK=[UInt8](repeating:0,count:DILITHIUM.PK_SIZE_3)
        var SIG=[UInt8](repeating:0,count:DILITHIUM.SIG_SIZE_3)

	    var rng=RAND()
		rng.clean()
	    for i in 0..<100 {RAW[i]=UInt8((i+1)&0xff)}
	    rng.seed(100,RAW)

        print("\nTesting Dilithium Signature")
        var tats=0
        for _ in 0..<LOOPS {
            let mess="Hello World"
            for i in 0..<32 {
                TAU[i]=rng.getByte()
            }
            DILITHIUM.keypair_3(TAU,&SK,&PK)

/*
            for i in 0..<SK.count {
                let h=String(format: "%02x",SK[i])
            	print("\(h)", terminator: "")
        	}
            print("")
            for i in 0..<PK.count {
                let h=String(format: "%02x",PK[i])
            	print("\(h)", terminator: "")
        	}
            print("")
*/

            let attempts=DILITHIUM.signature_3(SK,[UInt8](mess.utf8),&SIG)
            tats += attempts
            let h=String(format: "Signature %d bits created on attempt %d\n",8*SIG.count,attempts)
            print("\(h)", terminator: "") 

            let result=DILITHIUM.verify_3(PK,[UInt8](mess.utf8),SIG)
            if result {
                print("Signature is verified")
            } else {
                print("Signature is NOT verified")
                break
            }

        }
        if LOOPS > 1 {
            let h=String(format: "Average = %d\n",tats/LOOPS)
            print("\(h)", terminator: "") 
        }
}

TestDLTHM()

File: java/TestDLTHM.java
/*
 * Copyright (c) 2012-2021 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* test driver and function exerciser for Dilithium functions */

// See https://eprint.iacr.org/2017/633 (Ducas, Lepoint, Lyubashevsky, Schwabe, Seiler, Stehle)

package org.miracl.core;

import java.util.Scanner;
import junit.framework.TestCase;       //
import org.miracl.core.RAND;
import org.miracl.core.DILITHIUM;

public class TestDLTHM extends TestCase {
    public static void testDLTHM() {
        int i,attempts,LOOPS=100;
        boolean result;
        byte[] RAW = new byte[100];
        byte[] TAU = new byte[32];
		byte[] SK= new byte[DILITHIUM.SK_SIZE_3];
		byte[] PK= new byte[DILITHIUM.PK_SIZE_3];
		byte[] SIG=new byte[DILITHIUM.SIG_SIZE_3];
		//byte[] CT=new byte[KYBER.CIPHERTEXT_SIZE_768];
        //byte[] M = new byte[128];
    
        RAND RNG = new RAND();
      
        for (i = 0; i < 100; i++) RAW[i] = (byte)(i + 1);
        RNG.seed(100, RAW);

        System.out.println("\nTesting Dilithium signature");

		int tats=0;
        for (int j=0;j<LOOPS;j++) {
			String mess=new String("Hello World");
			for (i = 0; i < 32; i++) TAU[i] = (byte)RNG.getByte(); //(byte)(i + 1);
			
            DILITHIUM.keypair_3(TAU,SK,PK);
/*
            System.out.printf("SK= 0x");
            for (i = 0; i < SK.length; i++)
                System.out.printf("%02x", SK[i]);
            System.out.println();

            System.out.printf("PK= 0x");
            for (i = 0; i < PK.length; i++)
                System.out.printf("%02x", PK[i]);
            System.out.println();
*/
            attempts=DILITHIUM.signature_3(false,null,SK,null,mess.getBytes(),SIG);
            tats+=attempts;
/*
            System.out.printf("Signature = 0x");
            for (i = 0; i < SIG.length; i++)
                System.out.printf("%02x", SIG[i]);
            System.out.println();
*/
            System.out.printf("Signature %d bits created on attempt %d\n",8*SIG.length,attempts);

            result=DILITHIUM.verify_3(false,PK,null,mess.getBytes(),SIG);
            if (result) {
                System.out.printf("Signature is verified\n");
            } else {
                System.out.printf("Signature is NOT verified\n"); 
                break;
            }
		}
        if (LOOPS>1)
            System.out.printf("Average= %d\n",tats/LOOPS);
	}
}
